### DISCUSSION TITLE
Removing objects from D-bus, changing their interfaces

### DISCUSSION THREAD

**Discussion by dragomirecky** [python-sdbus/python-sdbus#57] [General]
Hi,

First of all, nice library! Thanks!

I am trying it on a new project and I might be missing two features. So I would like to discuss what is your plan regarding those and if I might be of help somehow.

First: Is there a way to unexport an object from D-bus? The only thing I found is that it gets removed when the object is removed by the GC. However, that is hard-to-work with and it is quite nondeterministic. 

Second (and lower priority to me than the first one): What about adding & removing interfaces of an exported object? Have you how this would be done with the current api?

And once again, thanks for the library ☺️

**Comment by igo95862** [comment]
Hello @dragomirecky 

> First: Is there a way to unexport an object from D-bus? The only thing I found is that it gets removed when the object is removed by the GC. However, that is hard-to-work with and it is quite nondeterministic.

It is actually quite deterministic if you don't create circular references. Remember that CPython uses the reference counting as its primary mechanism for resource control. Once all references get dropped the object will disappear from D-Bus. The garbage collector only comes in play when a circular reference was created. (for example list that contains it self)

However, I might add some kind of call to guaranteed remove object from exported objects.

> What about adding & removing interfaces of an exported object? Have you how this would be done with the current api?

The sd-bus allows exporting new vtables under same path if they don't share the interface names. So it might already be possible with the current API as long as the interfaces are different. (meaning classes are not related)

 However, is there a particular design pattern that can use objects adding and removing interfaces?

**Comment by dragomirecky** [comment]
> > First: Is there a way to unexport an object from D-bus? The only thing I found is that it gets removed when the object is removed by the GC. However, that is hard-to-work with and it is quite nondeterministic.
> 
> It is actually quite deterministic if you don't create circular references. Remember that CPython uses the reference counting as its primary mechanism for resource control. Once all references get dropped the object will disappear from D-Bus. The garbage collector only comes in play when a circular reference was created. (for example list that contains it self)
> 
> However, I might add some kind of call to guaranteed remove object from exported objects.

Yep, you are right and I am aware of that. However, still such an API is a bit clumsy to use when there is a need to have fine-control of when the object disappears from the D-bus. Or be able to remove an object from D-bus without deallocating it (and being able to reexport it later). I would welcome something like remove_from_dbus() method. In case we would agree on the API, I might create a PR with that.

> > What about adding & removing interfaces of an exported object? Have you how this would be done with the current api?
> 
> The sd-bus allows exporting new vtables under same path if they don't share the interface names. So it might already be possible with the current API as long as the interfaces are different. (meaning classes are not related)

Thanks, that is good to know and that solves my problem 👍

> However, is there a particular design pattern that can use objects adding and removing interfaces?

From reading the D-bus documentation I got the impression that generally, the granularity at which things are added/removed to D-bus are interfaces. For example the ObjectManager interface, which has InterfaceAdded/Removed signals. And it would (maybe) fit one of my project's needs to dynamically expose additional interface on an object, when I find out it has another feature available. But feel free to correct me if I am getting it wrong, I am new to D-bus

**Reply by igo95862** [reply]
> I would welcome something like remove_from_dbus() method. 

I am thinking about a context manager at the moment:

```python
with dbus_object.export_to_dbus("/"):
    ...
```

Maybe `remove_export` as well. Not sure if it better to be a object method or a bus method.

> For example the ObjectManager interface, which has InterfaceAdded/Removed signals.

I believe it is to be able to determine the type of an object that has been added. For example, in NetworkManager that makes use of ObjectManager the network devices have 2 common interfaces and 1 device type specific. So by comparing the interfaces of an device type you can determine the device type. This is also how [`sdbus.utils.parse_interfaces_added`](https://python-sdbus.readthedocs.io/en/latest/utils.html#sdbus.utils.parse_interfaces_added) works.

**Reply by dragomirecky** [reply]
> > I would welcome something like remove_from_dbus() method.
> 
> I am thinking about a context manager at the moment:
> 
> ```python
> with dbus_object.export_to_dbus("/"):
>     ...
> ```
> 
> Maybe `remove_export` as well. Not sure if it better to be a object method or a bus method.

Having a context manager is definitely a good idea. So maybe something like the `export_to_dbus()` returning an "dbus export" object having `remove()` + being a context manager.

**Reply by igo95862** [reply]
Hello @dragomirecky 

Would you want such functionality in upcoming 0.12.0 release or a later one? I will probably release a 0.12.rc1 next weekend.

The ObjectManager's `export_with_manager` will also be changed to return a handle.

**Reply by dragomirecky** [reply]
Well, if possible, it would be awesome having them in the 0.12.0. But no rush. And feel free to tell me if I can help with something :)

**Reply by igo95862** [reply]
It is a simple implementation.

> And feel free to tell me if I can help with something

More feedback is always welcomed. The current development tree has some big reworks compared to the version 0.11.1. I am worried were are some hidden regressions.

**Reply by igo95862** [reply]
@dragomirecky I added the handle in commit 331b6e91296490a280b3bc5510f263a12b24c39c.

[Documentation has a few examples how it can be used.](https://python-sdbus.readthedocs.io/en/latest/asyncio_api.html#sdbus.DbusInterfaceCommonAsync.export_to_dbus)

I plan on releasing the 0.12.rc1 tomorrow.

**Comment by igo95862** [comment]
0.12.RC1 has been released which includes the handle: https://github.com/python-sdbus/python-sdbus/releases/tag/0.12.RC1

**Reply by igo95862** [reply]
0.12.0 has been released.

**Reply by dragomirecky** [reply]
looks great! Thanks a lot 🖤
